exonum/src/helpers/fabric/password.rs

```diff
@@ -0,0 +1,164 @@
+
+use helpers::ZeroizeOnDrop;
+
+/// Passphrase input methods
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum PassInputMethod {
+    /// Prompt passphrase from terminal.
+    Terminal,
+    /// Get passphrase from environment variable (default if `None`).
+    EnvVariable(Option<String>),
+    /// Passphrase is passed as a command line parameter.
+    CmdLineParameter(ZeroizeOnDrop<String>),
+}
+
+/// Secret key types.
+#[derive(Copy, Clone)]
+pub enum SecretKeyType {
+    Consensus,
+    Service,
+}
+
+impl PassInputMethod {
+    /// Get passphrase using selected method.
+    /// Details of this process differs for different secret key types and whether we run node
+    /// or generate config files.
+    pub fn get_passphrase(self, key_type: SecretKeyType, node_run: bool) -> ZeroizeOnDrop<String> {
+        match self {
+            PassInputMethod::Terminal => {
+                let prompt = match key_type {
+                    SecretKeyType::Consensus => "Enter consensus key passphrase",
+                    SecretKeyType::Service => "Enter service key passphrase",
+                };
+                prompt_passphrase(prompt, node_run).expect("Failed to read password from stdin")
+            }
+            PassInputMethod::EnvVariable(name) => {
+                let var = if let Some(ref name) = name {
+                    name
+                } else {
+                    match key_type {
+                        SecretKeyType::Consensus => "EXONUM_CONSENSUS_PASS",
+                        SecretKeyType::Service => "EXONUM_SERVICE_PASS",
+                    }
+                };
+                ZeroizeOnDrop(env::var(var).unwrap_or_else(|e| {
+                    panic!("Failed to get password from env variable: {}, {}", var, e)
+                }))
+            }
+            PassInputMethod::CmdLineParameter(pass) => pass,
+        }
+    }
+}
+
+impl Default for PassInputMethod {
+    fn default() -> Self {
+        PassInputMethod::Terminal
+    }
+}
+
+impl FromStr for PassInputMethod {
+    type Err = failure::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "" {
+            return Ok(Default::default());
+        }
+
+        if s == "stdin" {
+            return Ok(PassInputMethod::Terminal);
+        }
+
+        if s.starts_with("env") {
+            let env_var = s.split(':').nth(1).map(|s| s.to_owned());
+            return Ok(PassInputMethod::EnvVariable(env_var));
+        }
+
+        if s.starts_with("pass") {
+            let pass = s.split(':').nth(1).unwrap_or("");
+            return Ok(PassInputMethod::CmdLineParameter(ZeroizeOnDrop(
+                pass.to_owned(),
+            )));
+        }
+
+        bail!("Failed to parse passphrase input method");
+    }
+}
+
+fn prompt_passphrase(prompt: &str, node_run: bool) -> io::Result<ZeroizeOnDrop<String>> {
+    if node_run {
+        return Ok(ZeroizeOnDrop(read_password_from_tty(Some(prompt))?));
```

Comment:
Can we use an empty passphrase when running node?

Reply:
Yes. That is why this code is here, if we generated key files without password we should be able to use them using all methods of entering passphrase including terminal.

Reply:
How can we generate files without password? Using `run-dev`? According to this method, we can't enter an empty password from the terminal of maybe I missing something.

Reply:
Using run dev.
Using "no password option".
Using empty password in environment variables.